storm-server/src/main/java/org/apache/storm/scheduler/blacklist/strategies/RasBlacklistStrategy.java

```diff
@@ -79,25 +81,46 @@
 
             if (shortage.areAnyOverZero() || shortageSlots > 0) {
                 LOG.info("Need {} and {} slots more. Releasing some blacklisted nodes to cover it.", shortage, shortageSlots);
-                //release earliest blacklist
-                for (String supervisor : blacklistedNodeIds) {
-                    SupervisorDetails sd = availableSupervisors.get(supervisor);
-                    if (sd != null) {
-                        NormalizedResourcesWithMemory sdAvailable = cluster.getAvailableResources(sd);
-                        int sdAvailableSlots = cluster.getAvailablePorts(sd).size();
-                        readyToRemove.add(supervisor);
-                        shortage.remove(sdAvailable, cluster.getResourceMetrics());
-                        shortageSlots -= sdAvailableSlots;
-                        LOG.debug("Releasing {} with {} and {} slots leaving {} and {} slots to go", supervisor,
-                            sdAvailable, sdAvailableSlots, shortage, shortageSlots);
-                        if (!shortage.areAnyOverZero() && shortageSlots <= 0) {
-                            // we have enough resources now...
-                            break;
+
+                //release earliest blacklist - but release all supervisors on a given blacklisted host.
+                Map<String, Set<String>> hostToSupervisorIds = createHostToSupervisorMap(blacklistedNodeIds, cluster);
+                for (Set<String> supervisorIds : hostToSupervisorIds.values()) {
+                    for (String supervisorId : supervisorIds) {
+                        SupervisorDetails sd = availableSupervisors.get(supervisorId);
+                        if (sd != null) {
+                            NormalizedResourcesWithMemory sdAvailable = cluster.getAvailableResources(sd);
+                            int sdAvailableSlots = cluster.getAvailablePorts(sd).size();
+                            readyToRemove.add(supervisorId);
+                            shortage.remove(sdAvailable, cluster.getResourceMetrics());
+                            shortageSlots -= sdAvailableSlots;
+                            LOG.info("Releasing {} with {} and {} slots leaving {} and {} slots to go", supervisorId,
+                                    sdAvailable, sdAvailableSlots, shortage, shortageSlots);
```

Comment:
So what is the purpose we need to release supervisors blacklist grouping by nodes. What if the node is stuck or broken ? If we shuffle the release by nodes, may be the risk would be spread evenly.

Reply:
If there are two supervisors on a host that are blacklisted and we release one, one supervisor will remain blacklisted.  This single blacklisted supervisor will be returned here:

https://github.com/apache/storm/blob/master/storm-server/src/main/java/org/apache/storm/scheduler/blacklist/BlacklistScheduler.java#L167

Then this code will see that the same host that both supervisors are on is blacklisted, preventing any scheduling on that node:

https://github.com/apache/storm/blob/master/storm-server/src/main/java/org/apache/storm/scheduler/blacklist/BlacklistScheduler.java#L167

I think a better overall long term solution is to blacklist supervisors instead of hosts, but that touches a lot more code.  In the short term I think this is a relatively small tradeoff to allow blacklisting to work with multiple supervisors per host.

Reply:
@agresch Got your idea.
In my personal opinion, for production, if a node has a blacklist supervisor(which means the supervisor does not send any HBs for some time), most of the cases it because the node machine itself has some problems(for now there are a few causes like: disk is full or network is in disconnection), so a safer way is we never schedule workers to the node if there are some blacklist supervisors on it.

If you want to make use of the healthy supervisor on the node(has some blacklist supervisors also), at least there is a decision to make sure the supervisor is healthy, we can do this through checking the heartbeats of it.

Reply:
It sounds like you don't agree even with the existing code that releases blacklisted supervisors when needed for scheduling? 

Should I be looking at converting all the existing code to blacklist supervisors instead of hosts to get things functional when there are more than one supervisor on a host?

Thanks.